master/internal/logpattern/logpattern.go

```diff
@@ -0,0 +1,333 @@
+package logpattern
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"sync"
+	"testing"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/task/tasklogger"
+	"github.com/determined-ai/determined/master/internal/webhooks"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+	"github.com/determined-ai/determined/master/pkg/set"
+
+	"github.com/uptrace/bun"
+)
+
+var (
+	blockListCache = make(map[model.TaskID]*set.Set[string])
+	mu             sync.RWMutex
+	regexCache     *lru.Cache[string, *regexp.Regexp]
+	regexCacheSize = 128
+)
+
+func webhookTypeFromPolicy(p expconf.SendWebhookPolicy) webhooks.WebhookType {
+	switch p.WebhookType() {
+	case "default":
+		return webhooks.WebhookTypeDefault
+	case "slack":
+		return webhooks.WebhookTypeSlack
+	default:
+		return webhooks.WebhookTypeDefault
+	}
+}
+
+// Monitor checks for logs against any log_pattern_policies and takes action according to the policy.
+func Monitor(ctx context.Context,
+	taskID model.TaskID, logs []*model.TaskLog, policies expconf.LogPatternPoliciesConfig,
+) error {
+	if len(policies) == 0 {
+		return nil
+	}
+
+	// TODO when we add rm specific log grabbing we will need to also monitor them.
+	for _, l := range logs {
+		if l.AgentID == nil {
+			return fmt.Errorf("agentID must be non nil to monitor logs")
+		}
+
+		for _, lpp := range policies {
+			// TODO we have this problem where a regex will always match itself.
```

Comment:
Do you this this is an issue we should try and solve?

Reply:
I didn't notice this. I have some ideas. I will fix this.